### DISCUSSION TITLE
how to use column value as index to a list in a dictionary.

### DISCUSSION THREAD

**Discussion by Hasham04** [vaexio/vaex#1925] [Q&A]
I have a dictionary like this

dictionary={somekey:[v1,v2,v3,v4,v5,v6,v7]}

I want to assign values to a column based on the value of another column.

col1        col2
1            v2
1            v2
1            v2 
1            v2

here 1 in col1 represents dictionary[somekey][1]

I have tried this.

df['col2']=df.func.where(df.col1=='1' #example,dictionary['somekey'][int(df.col1)],'')

this says the index of a dictionary cant be an expression which makes sense because columns are expressions, but still, need a way to do this.

Thanks in advance for any help.

**Comment by JovanVeljanoski** [comment]
I really don't understand what you are trying to do. Can you also use code formatting please?
Would the `map` method help?

**Reply by Hasham04** [reply]
I am trying to look up the value of a column as an index in a dictionary key that has a list as a value. if the values of the column are 1,2,3,4. I want the values at index 1,2,3,4 as the values.

**Reply by JovanVeljanoski** [reply]
if I understand correctly.. `df.col1.map(your_mapping_dictionary)` should work? Or do i still misunderstand?

**Reply by Hasham04** [reply]
how do I convert the values of two columns to a dictionary?

col1 col2
1      2
2      3
3      4

dictionary={1:2, 2:3, 3:4} like this.

**Reply by Hasham04** [reply]
`would  dict(zip(col1,col2)) work
`

**Reply by JovanVeljanoski** [reply]
The simplest way I know is something like this example

```python
import vaex

df = vaex.datasets.titanic()
tmp = df[['name', 'age']].to_pandas_df()
tmp = tmp.set_index('name')
mapping = dict(tmp)
print(mapping)
```

It is best if you illustrate your problem in a fully reproducible example like i've done before. 
This is the only way I can be sure of what your intent is. 
Not sure if this is the best approach btw, and you are putting things in memory by calling pandas.
Maybe if you give me a full example, i can play with it (if it is simple).

**Reply by Hasham04** [reply]
```
import vaex
df = vaex.datasets.titanic()
tmp = df['name', 'age'].to_pandas_df()
tmp = tmp.set_index('name')['age'].to_dict()
```
is about 20x faster

is there any faster non-ram using way to create a lookup that I can pass to `.map `, or any technique that will less memory when creating the lookup or when using `.map`?

Any suggestions would be helpful. Thankyou.

Second question how do you feel about people using vaex for complex transformations of large datasets instead of pandas. I am trying to figure out creative ways to implement what `groupby` with a custom function and `explode` do in pandas in vaex hahaha.

**Reply by JovanVeljanoski** [reply]
Wait,  if  both of your columns are in the dataframe you want to transform.. what is the point of the thing you are trying to do?  Or do you have two dataframes and trying to map something from one to another?

For a single dataframe, `df.func.where` should take you where want to go (i know you mentioned it above, but the use is incorrect). If you have to dataframes, then maybe some trick with join will do. 

Again, maybe a more complete example illustrating what you are trying to do might help.

As for your question: vaex is built for transforming big data, and for processes that pandas is too slow or the memory usage is too high. As for custom aggregations and explode.. those are tricky to do (which is why we have not done them). I think @maartenbreddels had ideas on how to approach them. But feel free to open a PR!

**Reply by Hasham04** [reply]
```
Id  | Col in final file | Val
1.  | Final_id.           |.  Val for col in Final file for Id=1
1. |. Col 2.                |.  Val for col in final file for Id=1
1. |. Col 2.                | Val for col in final file for Id=1

```
Multiple values for col2 in the final file can have the same final_id val. A one to many relationship. So I'm trying remember what that will be by mapping` {Id : value for final_id}`. Right now I am applying `.map` to a new column which contains Ids  if Col in final file is Col2 otherwise "" by using `.where`. which I am then filtering out "" to only contains Id if it is col2 in the final file this removes any final_ids. I then use the map to apply which final_id value col2 belongs to on this new column. Phew. I really hope this makes sense .

**Reply by Hasham04** [reply]
Please let me know if I am still unclear. I appreciate your help.

**Reply by JovanVeljanoski** [reply]
I am sorry I want to help but I don't understand your case. Maybe somebody else understands it better.

**Reply by Hasham04** [reply]
Appreciate it regardless. Thanks you. I will play around with joins and post my findings here.

**Reply by maartenbreddels** [reply]
This would be a pure Python thing:
```
dict(zip(df.x.tolist(), df.y.tolist()))
```
But then I don't see a reason to use Vaex

**Reply by Hasham04** [reply]
hey Maarten. I was able to accomplish this using joins. thankyou for your response.
